skxray/img_proc/mathops.py

```diff
@@ -0,0 +1,349 @@
+
+        In the event that bitwise operations are required the operators &,
+        |, ^, ~ may also be used, though I'm struggling to come up with a
+        scenario where this will be used.
+
+        Order of operations and parenthesis are taken into account when
+        evaluating the expression.
+
+    A : {ndarray, int, float}
+        Data set or constant to be offset or manipulated
+
+    B : {ndarray, int, float}
+        Data set or constant to be offset or manipulated
+
+    C : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+    D : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+    E : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+    F : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+    G : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+    H : {ndarray, int, float}, optional
+        Data set or constant to be offset or manipulated
+
+
+    Returns
+    -------
+    output : {ndarray, int, float}
+        Returns the resulting array or value to the designated variable
+
+    Example
+    -------
+    result = mathops.arithmetic_custom('(A+C)/(B+D)', img_1, img_2, 2, 4)
+    """
+    output = eval(parser.expr(expression).compile())
+    return output
+
+
+def logic_basic(operation,
+                src_data1,
+                src_data2=None):
+    """
+    This function enables the computation of the basic logical operations
+    oft used in image processing of two image or volume  data sets. This
+    function can be used for data comparison, material isolation,
+    noise removal, or mask application/generation.
+
+    Parameters
+    ----------
+    operation : str
+        options include:
+            'and' -- 2 inputs
+            'or' -- 2 inputs
+            'not' -- 1 input
+            'xor' -- 2 inputs
+            'nand' -- 2 inputs
+            'subtract' -- 2 inputs
+
+    src_data1 : {ndarray, int, float, list, tuple}
+        Specifies the first reference
+
+    src_data2 : {ndarray, int, float, list, tuple}
+        Specifies the second reference
+
+    Returns
+    -------
+    output : {ndarray, bool}
+        Returns the result of the logical operation, which can be an array,
+        or a simple boolean result.
+
+    Example
+    -------
+    result = mathops.logic_basic('and', img_1, img_2)
+    """
+    logic_dict = {'and' : np.logical_and,
+                  'or' : np.logical_or,
+                  'not' : np.logical_not,
+                  'xor' : np.logical_xor,
+                  'nand' : logical_NAND,
+                  'nor' : logical_NOR,
+                  'subtract' : logical_SUB
+                  }
+    output = logic_dict[operation](src_data1,
+                                   src_data2)
+    return output
+
+
+def logical_NAND(src_data1,

```

Comment:
I would also suggest making the docstrings for these functions identical to their cousins in numpy:

```python
ef logical_and(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_and(x1, x2[, out])
    
    Compute the truth value of x1 AND x2 element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Input arrays. `x1` and `x2` must be of the same shape.
    
    
    Returns
    -------
    y : {ndarray, bool}
        Boolean result with the same shape as `x1` and `x2` of the logical
        AND operation on corresponding elements of `x1` and `x2`.
    
    See Also
    --------
    logical_or, logical_not, logical_xor
    bitwise_and
    
    Examples
    --------
    >>> np.logical_and(True, False)
    False
    >>> np.logical_and([True, False], [False, False])
    array([False, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_and(x>1, x<4)
    array([False, False,  True,  True, False], dtype=bool)
    """
    pass

def logical_not(x, out=None): # real signature unknown; restored from __doc__
    """
    logical_not(x[, out])
    
    Compute the truth value of NOT x element-wise.
    
    Parameters
    ----------
    x : array_like
        Logical NOT is applied to the elements of `x`.
    
    Returns
    -------
    y : bool or ndarray of bool
        Boolean result with the same shape as `x` of the NOT operation
        on elements of `x`.
    
    See Also
    --------
    logical_and, logical_or, logical_xor
    
    Examples
    --------
    >>> np.logical_not(3)
    False
    >>> np.logical_not([True, False, 0, 1])
    array([False,  True,  True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_not(x<3)
    array([False, False, False,  True,  True], dtype=bool)
    """
    pass

def logical_or(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_or(x1, x2[, out])
    
    Compute the truth value of x1 OR x2 element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Logical OR is applied to the elements of `x1` and `x2`.
        They have to be of the same shape.
    
    Returns
    -------
    y : {ndarray, bool}
        Boolean result with the same shape as `x1` and `x2` of the logical
        OR operation on elements of `x1` and `x2`.
    
    See Also
    --------
    logical_and, logical_not, logical_xor
    bitwise_or
    
    Examples
    --------
    >>> np.logical_or(True, False)
    True
    >>> np.logical_or([True, False], [False, False])
    array([ True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_or(x < 1, x > 3)
    array([ True, False, False, False,  True], dtype=bool)
    """
    pass

def logical_xor(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_xor(x1, x2[, out])
    
    Compute the truth value of x1 XOR x2, element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Logical XOR is applied to the elements of `x1` and `x2`.  They must
        be broadcastable to the same shape.
    
    Returns
    -------
    y : bool or ndarray of bool
        Boolean result of the logical XOR operation applied to the elements
        of `x1` and `x2`; the shape is determined by whether or not
        broadcasting of one or both arrays was required.
    
    See Also
    --------
    logical_and, logical_or, logical_not, bitwise_xor
    
    Examples
    --------
    >>> np.logical_xor(True, False)
    True
    >>> np.logical_xor([True, True, False, False], [True, False, True, False])
    array([False,  True,  True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_xor(x < 1, x > 3)
    array([ True, False, False, False,  True], dtype=bool)
    
    Simple example showing support of broadcasting
    
    >>> np.logical_xor(0, np.eye(2))
    array([[ True, False],
           [False,  True]], dtype=bool)
    """
    pass
```